shell/server/proxy.js

```diff
@@ -578,29 +581,69 @@ function gcSessions() {
 }
 Meteor.setInterval(gcSessions, 60000);
 
-// Try to restore sessions on server restart.
-Meteor.startup(function () {
-  // Delete stale sessions from session list.
-  gcSessions();
-
-  // Remake proxies for all sessions that remain.
-  Sessions.find({}).forEach(function (session) {
-    var grain = Grains.findOne(session.grainId);
-    if (!grain) return;
-
-    var user = null;
-    if (session.userId) {
-      user = Meteor.users.findOne({_id: session.userId});
-      if (!user) return;  // Session owner no longer exists.
-    }
+var getProxyForHostId = function (hostId) {
+  // Get the Proxy corresponding to the given grain session host, possibly (re)creating it if it
+  // doesn't already exist. The first request on the session host will always create a new proxy.
+  // Later requests may create a proxy if they go to a different front-end replica or if the
+  // front-end was restarted.
+  check(hostId, String);
+
+  return Promise.resolve(undefined).then(function () {
+    var proxy = proxiesByHostId[hostId]
+    if (proxy) {
+      return proxy;
+    } else {
+      // Set table entry to null for now so that we can detect if it is concurrently deleted.
+      proxiesByHostId[hostId] = null;
+
+      return inMeteor(function () {
+        var session = Sessions.findOne({hostId: hostId});
+        if (!session) {
+          // Does not appear to be a valid session host.
+          return undefined;
+        }
 
-    var isOwner = grain.userId === session.userId;
-    var proxy = new Proxy(session.grainId, grain.userId, session._id, session.hostId, isOwner,
-                          user, null, false);
-    proxies[session._id] = proxy;
-    proxiesByHostId[session.hostId] = proxy;
+        var apiToken;
+        if (session.hashedToken) {
+          apiToken = ApiTokens.findOne({_id: session.hashedToken});
+          // We don't have to fully validate the API token here because if it changed the session
+          // would have been deleted.
+          if (!apiToken) {
+            throw new Meteor.Error(410, "ApiToken has been deleted");
+          }
+        }
+
+        var grain = Grains.findOne(session.grainId);
+        if (!grain) {
+          // Grain was deleted, I guess.
+          throw new Meteor.Error(410, "Resource has been deleted");
+        }
+
+        // Note that we don't need to call mayOpenGrain() because the existence of a session
+        // implies this check was already performed.
+        //
+        // TODO(perf): The Proxy is going to recompute the user's permissions. Maybe those should
+        //   be cached in the Session?
```

Comment:
To be clear: these values are already written to `Sessions`, but we currently do *not* take advantage of that fact to avoid repeated permissions computations.